core/eth/operatorstate/operator_state_cache.go

```diff
@@ -0,0 +1,144 @@
+package operatorstate
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Layr-Labs/eigenda/common"
+	"github.com/Layr-Labs/eigenda/core"
+	"github.com/Layr-Labs/eigenda/core/eth"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+// the size of the index lock used by the OperatorStateCache
+const indexLockSize = 64
+
+// Responsible for fetching and caching operator state for a given reference block number and quorums.
+//
+// This utility is thread safe, and can be used in performance sensitive, multithreaded environments.
+type OperatorStateCache interface {
+	// GetOperatorState retrieves the operator state for a given reference block number and quorums
+	GetOperatorState(
+		ctx context.Context,
+		referenceBlockNumber uint64,
+		quorums []core.QuorumID,
+	) (*core.OperatorState, error)
+}
+
+var _ OperatorStateCache = (*operatorStateCache)(nil)
+
+// A standard implementation of the OperatorStateCache interface.
+type operatorStateCache struct {
+	// indexes chain data, required to get operator public keys
+	chainState core.ChainState
+
+	// used to get a list of quorums registered at a given reference block number
+	quorumScanner eth.QuorumScanner
+
+	// A cache for operator state, indexed by reference block number.
+	// This cache implementation is thread safe.
+	cache *lru.Cache[uint64, *core.OperatorState]
+
+	// Used to prevent simultaneous lookup for a particular reference block number. Not used to protect data
+	// structures against concurrent access.
+	indexLock *common.IndexLock
+}
+
+// Create a new caching wrapper around ChainState for fetching operator state.
+func NewOperatorStateCache(
+	contractBackend bind.ContractBackend,
+	chainState core.ChainState,
+	registryCoordinatorAddress gethcommon.Address,
+	cacheSize uint64,
+) (OperatorStateCache, error) {
+
+	cache, err := lru.New[uint64, *core.OperatorState](int(cacheSize))
+	if err != nil {
+		return nil, fmt.Errorf("NewOperatorStateCache: %w", err)
+	}
+
+	qs, err := eth.NewQuorumScanner(contractBackend, registryCoordinatorAddress)
+	if err != nil {
+		return nil, fmt.Errorf("NewQuorumScanner: %w", err)
+	}
+
+	return &operatorStateCache{
+		chainState:    chainState,
+		quorumScanner: qs,
+		cache:         cache,
+		indexLock:     common.NewIndexLock(indexLockSize),
+	}, nil
+}
+
+// GetOperatorState retrieves the operator state for a given reference block number and quorums.
+func (c *operatorStateCache) GetOperatorState(
+	ctx context.Context,
+	referenceBlockNumber uint64,
+	quorums []core.QuorumID,
+) (*core.OperatorState, error) {
```

Comment:
So, if callers mutate the returned struct, it would mess up the values in the cache, right? Probably worth mentioning